working/2365 - primary constructors/feature-specification.md

```diff
@@ -0,0 +1,510 @@
+
+An `external` instance variable amounts to an `external` getter and an
+`external` setter. Such "variables" cannot be initialized by an
+initializing formal anyway, so they do not fit into the treatment implied
+by a primary constructor. Just use a normal declaration.
+
+```dart
+class ModifierClass {
+  late int x;
+  external double d;
+  ModifierClass(this.x);
+}
+
+class ModifierClass(this.x) {
+  late int x;
+  external double d;
+}
+```
+
+`ModifierClass` as written does not make sense (`x` does not have to be
+`late`), but there could be other constructors that do not initialize `x`.
+
+Super parameters can be declared in the same way as in non-primary
+constructors:
+
+```dart
+class A {
+  final int a;
+  A(this.a);
+}
+
+class B extends A {
+  B(super.a);
+}
+
+class A(final int a);
+class B(super.a) extends A;
+```
+
+Next, the constructor can be named, and it can be constant:
+
+```dart
+class Point {
+  final int x;
+  final int y;
+  const Point._(this.x, this.y);
+}
+
+class const Point._(final int x, final int y);
+```
+
+Note that the class header contains syntax that resembles the constructor
+declaration, which may be helpful when reading the code.
+
+The modifier `const` could have been placed on the class (`const class`)
+rather than on the class name. This proposal puts it on the class name
+because the notion of a "constant class" conflicts with with actual
+semantics: It is the constructor which is constant because it is able to be
+invoked during constant expression evaluation; it can also be invoked at
+run time, and there could be other (non-constant) constructors. This means
+that it is at least potentially confusing to say that it is a "constant
+class", but it is consistent with the rest of the language to say that this
+particular primary constructor is a "constant constructor". Hence `class
+const Name` rather than `const class Name`.
+
+The modifier `final` on a parameter in a primary constructor has no meaning
+for the parameter itself, because there is no scope where the parameter can
+be accessed. Hence, this modifier is used to specify that the instance
+variables declared by this primary constructor are `final`.
+
+In the case where the constructor is constant, and in the case where the
+declaration is an `inline` class or an `enum` declaration, the modifier
+`final` on every instance variable is required. Hence, it can be omitted:
+
+```dart
+inline class I {
+  final int x;
+  I.name(this.x);
+}
+
+inline class I.name(int x);
+
+class Point {
+  final int x;
+  final int y;
+  const Point(this.x, this.y);
+}
+
+class const Point(int x, int y);
+
+enum E {
+  one('a'),
+  two('b');
+
+  final String s;
+  const E(this.s);
+}
+
+enum E(String s) { one('a'), two('b') }
+```
+
+This mechanism follows an existing pattern, where `const` modifiers can be
+omitted in the case where the immediaty syntactic context implies that this
+modifier _must_ be present. For example `const [const C()]` can be written
```

Comment:
We had a somewhat similar dilemma with `=>` function declarations (not function literals) with declared return type `void`: Some errors are not reported (e.g., `void f() => 3;`) with `=>`, but they _are_ reported with the otherwise equivalent form using a class body (`void f() { return 3; }`). A rationale which has often been mentioned in order to justify this inconsistency is that the return type is right there with `=>` for the reader, but it may be out of sight when the function has a `{...}` body.

So we could use the same rationale to say that we don't have to put `const` just two tokens after `enum`, because everybody can see the `enum` when they can see the `const`; and yet, `const` has to be stated explicitly on body constructors because the `enum` keyword could be out of sight.

I would also be quite happy about allowing `const` to be omitted from body constructors of an enum declaration, based on the assumption that a reader of an enum declaration is likely to be aware of the fact that it is an enum declaration (e.g., because they differ visually from other declarations, due to the typically prominent list of values).